File: test/integration-tests/testset-rpc/cpp/function_provider.cpp
/********************************************************************
 * Copyright (c) Robert Bosch GmbH
 * All Rights Reserved.
 *
 * This file may not be distributed without the file ’license.txt’.
 * This file is subject to the terms and conditions defined in file
 * ’license.txt’, which is part of this source code package.
 *********************************************************************/

#include <csignal>
#include <iostream>
#include <memory>

#include "nlohmann/json.hpp"
#include "logging.hpp"
#include "mqtt_client.hpp"

using json = nlohmann::json;
using namespace iotea::core;

static const char SERVER_ADDRESS[] = "tcp://localhost:1883";
static const char FEATURE[] = "testable_talent";
static const char FUNC_ECHO[] = "echo";

class EchoTalent : public FunctionTalent {
   public:
    explicit EchoTalent(std::shared_ptr<Publisher> publisher)
        : FunctionTalent(FEATURE, publisher) {
        RegisterFunction(FUNC_ECHO, [](const json& args, const CallContext& context) {
            context.Reply(args[0]);
        });
        SkipCycleCheck(true);
    }
};

static std::shared_ptr<MqttClient> client = std::make_shared<MqttClient>(SERVER_ADDRESS, FEATURE);

void signal_handler(int signal) { client->Stop(); }

int main(int argc, char* argv[]) {
    auto talent = std::make_shared<EchoTalent>(client);
    client->RegisterTalent(talent);

    std::signal(SIGINT, signal_handler);

    client->Run();

    return 0;
}

File: test/integration-tests/testset-rpc/python/function_provider.py
##############################################################################
# Copyright (c) 2021 Bosch.IO GmbH
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
#
# SPDX-License-Identifier: MPL-2.0
##############################################################################

import asyncio
import json
import os
import logging
 
from iotea.core.talent_func import FunctionTalent
from iotea.core.logger import Logger
logging.setLoggerClass(Logger)
logging.getLogger().setLevel(logging.INFO)
os.environ['MQTT_TOPIC_NS'] = 'iotea/'


class FunctionProvider(FunctionTalent):
    def __init__(self, connection_string):
        super(FunctionProvider, self).__init__('function-provider-py', connection_string)

        # Register Functions
        self.register_function('echo', self.echo)

    async def echo(self, value, ev, evctx):
        self.logger.debug('Echo called')
        return value


async def main():
    function_provider = FunctionProvider('mqtt://localhost:1883')
    await function_provider.start()

LOOP = asyncio.get_event_loop()
LOOP.run_until_complete(main())
LOOP.close()
